Default non-HDF5 Exodus output even w/HDF5 enabled

roystgnr opened the pull request:
## Reason
Current Moose builds with HDF5 are a "roach motel" - you can check in (a new Exodus file) any time you like, but you can never leave (because you just unwittingly wrote an HDF5 file and now it won't work with non-HDF5 Moose builds).

## Design
The Moose Exodus class now has a "write_hdf5" parameter, defaulting to false, determining whether HDF5 should be written instead of old NetCDF3.

## Impact
Users who really want to write HDF5 output will now have to set this parameter.  (there may not be any of those - we really wanted *read* support for HDF5 but we don't yet have IGA Exodus *output* so there's no obvious use case for *write* support)

Refs #20033

Pinging @loganharbour and @cticenhour; sorry for the delay with this.

roystgnr committed:
Default non-HDF5 Exodus output even w/HDF5 enabled

Otherwise we tend to "leak" HDF5-requiring Exodus files into places like
gold files that are expected to still work with HDF5-disabled Moose
builds.

Refs #20033

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1015588/) on 0d7cb84 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20793/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20793/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 08d5cacc152ab726a7e9d57684093337d0910962`

loganharbour commented:
Maybe `write_hdf5` -> `hdf5` for the parameter? I think the write is implied.

We should also test the HDF5 write.

roystgnr commented:
> Maybe write_hdf5 -> hdf5 for the parameter? I think the write is implied.

It's implied by all the conversations we've had on Slack.  I don't think it's at all implied by the existence of a parameter named "hdf5" in the Exodus class.  I know if I was looking at the docs I'd just assume "of course I need to set hdf5=true to read hdf5".

> We should also test the HDF5 write.

Manually, I've already done so, by running `file` over the outputs of a `make test` and ensuring that they were now reported as NetCDF not HDF5.  Automatically ... I agree, in theory, but in practice I'm not sure how to do so and I wouldn't want figuring it out to hold up this PR.

roystgnr commented:
I can't even see a way to do it with exodiff - that'll let you demand 64-bit integers but it won't let you insist on HDF5 or on NetCDF3!?

roystgnr commented:
> Manually, I've already done so, by running file over the outputs of a make test and ensuring that they were now reported as NetCDF not HDF5.

Damn, or have I?  `file` seems to be willing to tell me "the truth but not the whole truth" here.  Let me see what else I can do.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1015872/) on 4b43fef wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20793/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1015880/) on 4b43fef wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>08d5ca</th>
      <th colspan=4>#20793 4b43fe</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.57%</td>
      <td>82.57%</td>
      <td>+0.00%</td>
      <td>83.33%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>74187</td>
      <td>74197</td>
      <td>+10</td>
      <td>10</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15660</td>
      <td>15660</td>
      <td>-</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20793/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20793/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20793/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 83.33% is less than the suggested 90.0%



This comment will be updated on new commits.

roystgnr commented:
Okay; got that figured out: I was trying `file` on an Exodus 8 mesh that actually wasn't written with hdf5, and file dutifully reported to me that it was NetCDF format.  Actual Exodus 8 HDF5 files are listed as `Hierarchical Data Format (version 5) data`, and newly written Moose outputs with this branch are NetCDF format.

I'm still not sure how to put a test in for this, though.  `file` is a standard command, but its outputs aren't standard.  On systems with `h5dump` installed we could make sure that it doesn't recognize an output file, but I don't want to demand that everyone install `h5dump`...

Right now RunCommand `processResults()` sets `self.success` if a command returns 0, `self.fail` for non-zero.  Could we use `self.skip` if a command returns 77 (the automake "skipped test" code)?  Or is it too late to use that anywhere outside `checkRunnable()`?

roystgnr committed:
Interpret exit code 77 as "skip"

This is the automake test suite convention, and it may be useful here
too.

roystgnr committed:
Test for inadvertent HDF5 if we can

I'm not sure how to do this on systems without h5dump (a standard HDF5
bin utility) installed, but this is better than nothing.

roystgnr commented:
Awesome.  Python subprocess.Popen behaves differently on different systems.

roystgnr committed:
Report a-posteriori-skip as SKIP

roystgnr commented:
Oh, or maybe Popen is behaving consistently but some systems are just lazier than others about making `/bin/sh` a strict Bourne shell rather than just a lax link to bash.

roystgnr committed:
Use exit not return; echo result first

In our CI some /bin/sh are pickier than others about how you leave the
shell, and regardless we should make it easier for people to diagnose
what happened.
